shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -22,6 +25,50 @@
 static NSString* const kPerformAction = @"TextInputClient.performAction";
 static NSString* const kMultilineInputType = @"TextInputType.multiline";
 
+static NSString* const kTextAffinityDownstream = @"TextAffinity.downstream";
+static NSString* const kTextAffinityUpstream = @"TextAffinity.upstream";
+
+static NSString* const kTextInputAction = @"inputAction";
+static NSString* const kTextInputType = @"inputType";
+static NSString* const kTextInputTypeName = @"name";
+
+static NSString* const kSelectionBaseKey = @"selectionBase";
+static NSString* const kSelectionExtentKey = @"selectionExtent";
+static NSString* const kSelectionAffinityKey = @"selectionAffinity";
+static NSString* const kSelectionIsDirectionalKey = @"selectionIsDirectional";
+static NSString* const kComposingBaseKey = @"composingBase";
+static NSString* const kComposingExtentKey = @"composingExtent";
+static NSString* const kTextKey = @"text";
+
+/**
+ * The affinity of the current cursor position. If the cursor is at a position representing
+ * a line break, the cursor may be drawn either at the end of the current line (upstream)
+ * or at the beginning of the next (downstream).
+ */
+typedef NS_ENUM(NSUInteger, FlutterTextAffinity) {
+  FlutterTextAffinityUpstream,
+  FlutterTextAffinityDownstream
+};
+
+/*
+ * Updates a range given base and extent fields.
+ */
+static flutter::TextRange UpdateRangeFromBaseExtent(NSNumber* base,
+                                                    NSNumber* extent,
+                                                    flutter::TextRange range) {
```

Comment:
Speaking of which, lets make this a const ref.

Reply:
Done and killed off the mutations.